core/corehttp/ipns_hostname.go

```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```

Comment:
On Sat, May 16, 2015 at 10:50:58AM -0700, Knut Ahlers wrote:
> Other solution which should also fix #1234 would be to change this
> to `/ipns/` path and let the IPNS stuff do the magic. (For IPNS
> there is already an exception for `Cache-Control`)

In #1233, the problem is:

  $ curl -IL swagger.luzifer.io

says “ah, I have a hostname, let me try to resolve it via a dnslink”,
and gets:

  /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

Then it attaches the URL path (empty or a slash, doesn't really
matter) and returns a 301 redirect to (modulo the old /ipfs/
injection):

  http://swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

When it serves *that* request, it says “ah, I have a hostname, let me
try to resolve it via a dnslink”, and then attaches the URL path (now
/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA) to redirect to:

  http://swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

which is:

  http://swagger.luzifer.io/{dnslink-from-second-call}/{dnslink-from-first-call}{empty-original-path}

What we're missing is a way to break out of this hostname-prefixing
loop.  I see a few choices here.

a. Defining any URL path to be an absolute path, so:

     $ curl -IL swagger.luzifer.io

   has an empty (enough) path, and would get resolved to
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA, but

     $ curl -IL swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

   has a non-empty path, so it would *also* resolve to
   ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA (and not need a
   redirect).

   But that makes life hard for folks who want pretty names without a
   base-58 multihash [1].

b. Defining any URL path to be relative, so:

     $ curl -IL swagger.luzifer.io

   would just serve up
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA without any
   redirects.  And

     $ curl -IL swagger.luzifer.io/index.html

   would serve up the content referenced by the ‘index.html’ link in
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA (again without
   redirects).

   But that makes life hard for folks to folks who want to use the
   gateway behind swagger.luzifer.io to access content outside of the
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA tree.

c. Come up with some way to signal whether or not your URL path is
   absolute.  So:

     $ curl -IL -H 'IPFS: absolute' swagger.luzifer.io/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA

   would give you the content of your absolute path, and:

     $ curl -IL -H 'IPFS: relative' swagger.luzifer.io/index.html

   would give you the content of
   /ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/index.html.

   But an IPFS header is pretty ugly ;).

I think the right choice depends on who's running the gateway.  Folks
running an IPFS gateway as a service to the community (or their
subnet) will likely want either (a) or (c), so they can hit arbitrary
IPFS resource.  Folks serving their own website out of an IPFS gateway
will likely want either (b) or (c) so they get pretty URLs (and maybe
so they don't shoulder the bandwidth of an open-to-all-IPFS-space
gateway).  Folks using browsers or writing clients will likely not
want (c), because setting headers can be awkward.  If we have a
smarter way to communicate the absolute/relative distinction, then I
think (c) is the best choice, but I'm not sure what that way would be.
Maybe another leading slash for absolute paths?

  swagger.luzifer.io//ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA
  swagger.luzifer.io/index.html

[1]: https://github.com/ipfs/go-ipfs/issues/1234#issuecomment-102674315